File: Merge_Sort/Merge_Sort.java
class Merge_Sort
{
    public static int temp[];

    // Conquer
    public static void conquer_merge(int[] array, int left, int right, int mid)
    {
        int k = left, i, j;

        for(i = left, j = mid + 1; i <= mid && j <= right; k++)
        {
            if(array[i] <= array[j])
            {
                temp[k] = array[i];
                i++;
            }
            else
            {
                temp[k] = array[j];
                j++;
            }
        }

        while(i <= mid)
        {
            temp[k] = array[i];
            i++;
            k++;
        }

        while(j <= right)
        {
            temp[k] = array[j];
            j++;
            k++;
        }

        for(i = left; i <= right; i++)
            array[i] = temp[i];
    }

    // Divide array into halves
    public static void divide(int[] array, int left, int right)
    {
        if(left < right)
        {
            int mid = left + (right - left) / 2;

            divide(array, left, mid);
            divide(array, mid + 1, right);

            conquer_merge(array, left, right, mid);
        }
    }

    public static void MergeSort(int array[], int size)
    {
        temp = new int[size];
        divide(array, 0, size - 1);
        temp = null;
    }

    // function ro print array
    public static void Print_Array(int[] array, int size)
    {
        for(int i = 0; i < size; i++)
            System.out.print(array[i] + " ");

        System.out.println();
    }

    public static void main(String[] args)
    {
        int[] array = {2, 4, 3, 1, 6, 8, 4};
        MergeSort(array, 7);
        Print_Array(array, 7);
    }
}

// Output
// 1 2 3 4 4 6 8

File: Merge_Sort/Merge_Sort.rb
$temp = []
# Conquer
def conquer_merge(array, left, right, mid)
    #temp = [None] * len(array)
    i = left
    j = mid + 1
    k = left

    while i <= mid and j <= right
        if array[i] <= array[j]
            $temp[k] = array[i]
            i += 1
        else
            $temp[k] = array[j]
            j += 1
        end
        k += 1
    end
    while i <= mid
        $temp[k] = array[i]
        i += 1
        k += 1
    end
    while j <= right
        $temp[k] = array[j]
        j += 1
        k += 1
    end
    while left <= right
        array[left] = $temp[left]
        left += 1
    end
end

# Divide array into halves
def divide(array, left, right)
    if left < right
        mid = (left + (right - left) / 2).to_i

        divide(array, left, mid)
        divide(array, mid + 1, right)

        conquer_merge(array, left, right, mid)
    end
end

def Merge_Sort(array)
    $temp = [0] * array.length
    divide(array, 0, array.length - 1)
    $temp.clear
end

array = [2, 4, 3, 1, 6, 8, 4] #Hard Coded Array

Merge_Sort(array)

print(array)

# Output
# 1 2 3 4 4 6 8
